pageserver/src/tenant_tasks.rs

```diff
@@ -0,0 +1,257 @@
+        "compaction loop stopped. State is {:?}",
+        tenant_mgr::get_tenant_state(tenantid)
+    );
+}
+
+static START_GC_LOOP: OnceCell<mpsc::Sender<ZTenantId>> = OnceCell::new();
+static START_COMPACTION_LOOP: OnceCell<mpsc::Sender<ZTenantId>> = OnceCell::new();
+
+/// Spawn a task that will periodically schedule garbage collection until
+/// the tenant becomes inactive. This should be called on tenant
+/// activation.
+pub fn start_gc_loop(tenantid: ZTenantId) -> anyhow::Result<()> {
+    START_GC_LOOP
+        .get()
+        .context("Failed to get START_GC_LOOP")?
+        .blocking_send(tenantid)
+        .context("Failed to send to START_GC_LOOP channel")?;
+    Ok(())
+}
+
+/// Spawn a task that will periodically schedule compaction until
+/// the tenant becomes inactive. This should be called on tenant
+/// activation.
+pub fn start_compaction_loop(tenantid: ZTenantId) -> anyhow::Result<()> {
+    START_COMPACTION_LOOP
+        .get()
+        .context("failed to get START_COMPACTION_LOOP")?
+        .blocking_send(tenantid)
+        .context("failed to send to START_COMPACTION_LOOP")?;
+    Ok(())
+}
+
+/// Spawn the TenantTaskManager
+/// This needs to be called before start_gc_loop or start_compaction_loop
+pub fn init_tenant_task_pool() -> anyhow::Result<()> {
+    let runtime = tokio::runtime::Builder::new_multi_thread()
+        .thread_name("tenant-task-worker")
+        .worker_threads(40) // Way more than necessary
+        .max_blocking_threads(100) // Way more than necessary
+        .enable_all()
+        .build()?;
+
+    let (gc_send, mut gc_recv) = mpsc::channel::<ZTenantId>(100);
+    START_GC_LOOP
+        .set(gc_send)
+        .expect("Failed to set START_GC_LOOP");
+
+    let (compaction_send, mut compaction_recv) = mpsc::channel::<ZTenantId>(100);
+    START_COMPACTION_LOOP
+        .set(compaction_send)
+        .expect("Failed to set START_COMPACTION_LOOP");
+
+    // TODO this is getting repetitive
+    let mut gc_loops = HashMap::<ZTenantId, watch::Sender<()>>::new();
+    let mut compaction_loops = HashMap::<ZTenantId, watch::Sender<()>>::new();
+
+    thread_mgr::spawn(
+        ThreadKind::TenantTaskManager,
+        None,
+        None,
+        "Tenant task manager main thread",
+        true,
+        move || {
+            runtime.block_on(async move {
+                let mut futures = FuturesUnordered::new();
+                loop {
+                    tokio::select! {
+                        _ = thread_mgr::shutdown_watcher() => {
+                            // TODO do this from thread_mgr? Extend it to work with tasks?
+                            for (_, cancel) in gc_loops.drain() {
+                                cancel.send(()).ok();
+                            }
+                            for (_, cancel) in compaction_loops.drain() {
+                                cancel.send(()).ok();
+                            }
+                            // TODO wait for tasks to die?
```

Comment:
I would say that this is important. For tasks like detach/delete it is desired to join threads to be sure that they no longer access the filesystem. It is needed to have this on per-tenant basis too. So it is possible to cancel task and join it. Cancellation in this case is implicitly performed via setting tenant state to stopping, so the lacking piece is something `join`-like for these tasks. I can imagine two possible solutions, store join handles somewhere globally, or have a special pair of channels so ne can send join request and block on receiving confirmation message. Example of the latter approach can be found here: https://github.com/neondatabase/neon/pull/1936/files#diff-0d4539a7193c24a7f3349df105b8bf31a2767e74bcd4898532cb7d44838bf499R439 

cc @SomeoneToIgnore wdyt?

Reply:
Also we should be asserting that "no gc/compaction" is running, rather than "no gc/compaction **threads/tasks** are running". People are free to call `gc_iteration` outside of `thread_mgr` and `tenant_tasks`, and that will mess up detach/delete.

Reply:
The sync/async boundary makes this so complicated. Neither of the two approaches suggested are easy:
1. Can't store join handles because the `FuturesUnordered` eats them
2. Can't use a channel because we could have multiple loops for a tenant and we want all of them to shut down

So I wrote something different in my [latest commit](https://github.com/neondatabase/neon/pull/1933/commits/bf76f43ea4c9d1cb0d5bcd47fafa12747fe98b36). I don't like it that much.. it's not very simple or readable, but if it's workable I'll polish it up a bit and we'll refine later. It has the benefit of also waiting for gc/compaction tasks that circumvent tenant_mgr or thread_mgr, for example [this one](https://github.com/neondatabase/neon/blob/tenant-tasks/pageserver/src/page_service.rs#L1071).

Reply:
One way to simplify, maybe: https://github.com/neondatabase/neon/compare/tenant-tasks...tenant-tasks-cleanup?expand=1

Reply:
>  It seems counter-intuitive to me that the "active -> idle" state transition would block on closing tenant tasks.

It shouldn't be a problem because active -> idle transition is not on the hot path, it happens in the background. It may be a problem if threads we're waiting for do not respond/are stuck. Then we cannot start the tenant again, or we can but it will be in a weird state when there is one semi-dead gc task and one that is new and trying to work.

> Can't store join handles because the FuturesUnordered eats them

Join handles can be stored separately without using FuturesUnordered like it is done in walreceiver.

> Can't use a channel because we could have multiple loops for a tenant and we want all of them to shut down

There should be only one gc/compaction loop for a tenant at a time, or I'm missing something? If we need to notify multiple consumers we can use something like tokio watch or broadcast channel.

Reply:
> Join handles can be stored separately without using FuturesUnordered like it is done in walreceiver.

How do we catch and report errors/panics from tasks if we don't await them? Catch unwind inside the task? Or is there some tokio settings to auto-report?

> There should be only one gc/compaction loop for a tenant at a time, or I'm missing something?

A few commits ago it was possible to have many loops per tenant because when we cancel any existing tasks [we don't wait](https://github.com/neondatabase/neon/blob/2b4c3cb9329046bdbbedf86ec1a2f3959d0c8b66/pageserver/src/tenant_tasks.rs#L183) for them to shutdown. I guess that is fixed now that we wait for shutdown when we go "active -> idle".

But anyway, there are two other problems I already mentioned:
1. We don't wait for gc/compaction that's not part of a thread/task
2. Race conditions between state transitions. Even if we solve these, there's another race condition in detach/delete. We only make sure that there are no gc/compactions at the start of the detach/delete, but gc/compaction can start during detach/delete. It could be possible to end up with a semi-deleted active tenant without noticing. We need to hold a lock (or equivalent) while deleting.

Reply:
> How do we catch and report errors/panics from tasks if we don't await them?

Hmm, you're right, it is possible to await on join handles, but this way you need to have some communication from the task itself because otherwise you wont know if it is finished or not so without awaiting this join handle will be kept forever, and FuturesUnordered solves this. I think walreceiver uses this message passing mechanism when spawned task communicates updates back to its parent. But here it is not needed, tasks are rather short lived, and this is another difference with walreceiver, where tasks live for whole compute lifetime. 

> Catch unwind inside the task? Or is there some tokio settings to auto-report?

Tokio will catch panic inside task, it uses catch unwind around the poll method of a future and dumps it into logs, but it wont increment metrics and so on. So you're right :)